grails-app/controllers/com/unifina/controller/api/ProductApiController.groovy

```diff
@@ -31,6 +36,24 @@ class ProductApiController {
 	ProductService productService
 	ProductImageService productImageService
 
+	@GrailsCompileStatic
+	@StreamrApi(allowRoles = AllowRole.ADMIN)
+	def staleProducts() {
+		List<Product> products = productService.list(new ProductListParams(publicAccess: true), loggedInUser())
+		int days = params.int("days", 2)
+		Date threshold = DateUtils.addDays(new Date(), -days)
+		List<ProductService.StaleProduct> results = productService.findStaleProducts(products, threshold)
+		System.out.println(results)
+		def json = results.collect { stale ->
+			[id: stale.product.id, name: stale.product.name, streams: stale.streams.collect { stream ->
+				StreamrMessage msg = stream.findMostRecentMessage()
```

Comment:
Aren't you already fetching the most recent message in `findStaleProducts`? Why not store it in the results to avoid having to look it up again?